File: code/cpp/eksperyment.cpp
// Eksperyment: porownanie czasu obliczania pierwiastka kwadratowego
// funkcja biblioteczna std::sqrt()  vs  wlasna metoda babilonska (dokladnosc 1e-15)
// Zakres: liczby calkowite 1..1000000, 5 powtorzen calej serii.
//
// Kompilacja: g++ -O2 -std=c++17 eksperyment.cpp -o eksperyment
// Uruchomienie: ./eksperyment > ../../experiment/wyniki_cpp.csv

#include <iostream>
#include <cmath>
#include <chrono>
#include <iomanip>

constexpr long N = 1000000;
constexpr int POWTORZENIA = 5;
constexpr double EPS = 1e-15;

double pierwiastekBabilonski(double S, double eps = EPS) {
    if (S <= 0) return 0.0;
    double x = S, poprzednie;
    do {
        poprzednie = x;
        x = 0.5 * (x + S / x);
    } while (std::fabs(x - poprzednie) > eps);
    return x;
}

int main() {
    using namespace std::chrono;

    std::cerr << std::fixed << std::setprecision(6);
    std::cout << "proba,sqrt_s,babilonska_s\n";

    for (int proba = 1; proba <= POWTORZENIA; ++proba) {
        // std::sqrt
        volatile double suma1 = 0.0;
        auto t1 = high_resolution_clock::now();
        for (long i = 1; i <= N; ++i) {
            suma1 += std::sqrt(static_cast<double>(i));
        }
        auto t2 = high_resolution_clock::now();
        double czas_sqrt = duration<double>(t2 - t1).count();

        // metoda babilonska
        volatile double suma2 = 0.0;
        auto t3 = high_resolution_clock::now();
        for (long i = 1; i <= N; ++i) {
            suma2 += pierwiastekBabilonski(static_cast<double>(i));
        }
        auto t4 = high_resolution_clock::now();
        double czas_bab = duration<double>(t4 - t3).count();

        std::cout << proba << "," << std::setprecision(6) << czas_sqrt
                  << "," << czas_bab << "\n";
        std::cerr << "Proba " << proba << ": sqrt=" << czas_sqrt
                  << "s  babilonska=" << czas_bab << "s"
                  << "  (suma kontrolna: " << suma1 << " / " << suma2 << ")\n";
    }
    return 0;
}

File: docs/code/python/eksperyment.py
"""Eksperyment (wersja Python, dodatkowa): sqrt() vs metoda babilonska.
Zakres 1..1_000_000, 5 powtorzen. Wynik zapisywany do CSV.
Uruchomienie: python3 eksperyment.py > ../../experiment/wyniki_python.csv
"""
import math
import time
import sys

N = 1_000_000
POWTORZENIA = 5
EPS = 1e-15


def pierwiastek_babilonski(s, eps=EPS):
    if s <= 0:
        return 0.0
    x = s
    while True:
        poprzednie = x
        x = 0.5 * (x + s / x)
        if abs(x - poprzednie) <= eps:
            return x


def main():
    print("proba,sqrt_s,babilonska_s")
    for proba in range(1, POWTORZENIA + 1):
        start = time.perf_counter()
        suma1 = sum(math.sqrt(i) for i in range(1, N + 1))
        czas_sqrt = time.perf_counter() - start

        start = time.perf_counter()
        suma2 = sum(pierwiastek_babilonski(i) for i in range(1, N + 1))
        czas_bab = time.perf_counter() - start

        print(f"{proba},{czas_sqrt:.6f},{czas_bab:.6f}")
        print(
            f"Proba {proba}: sqrt={czas_sqrt:.6f}s  babilonska={czas_bab:.6f}s "
            f"(suma kontrolna: {suma1:.3f} / {suma2:.3f})",
            file=sys.stderr,
        )


if __name__ == "__main__":
    main()
